Signals added to state machine

jedStevens opened the pull request:
I added signal emissions for when the AnimaiotnNodeStateMachinePlayback changes state. I'm not sure if I've really done this properly (in terms of style) so please feel free to suggest changes; but it does work as intended. 

Solves #22137

BastiaanOlij committed:
ARKit for Godot 3.2

This PR introduces support for ARKit to the iOS version of Godot.
ARKit is Apples Augmented Reality platform.
This PR brings in support for ARKit 1.0 and implements a few ARKit 2.0 features.
It requires iOS 11 to run but should not prevent Godot from running on older versions as long as ARKit remains unused.

Xrayez committed:
Add ability to expand/collapse nodes recursively in scene tree dock

To expand or collapse the node recursively (all children), hold `Shift`
button and click on the node's folding arrow.

The popup menu option `Expand/Collapse All" checks whether any node
is expanded or collapsed first and performs the opposite operation. That
means if any children node is collapsed, it will first expand all nodes
at selected node.

Co-authored-by: Rikhardur Bjarni Einarsson (MunWolf) badulf96@gmail.com

RandomShaper committed:
Skip export of non-project libraries

That is, any library referred to in GDNative library resources, won't be copied to the export target path unless its path begins with `res://`.

The case use for this is a bit advanced: having a GDN library that will be deployed separately from the project; for instance, to a path in the system (like `/opt/...`).

Currently the GDN library editor doesn't allow to pick dynamic libraries outside the project, but that can be done by editing the `.gdnlib` file manually.

AnaDenisa committed:
Add option to input value manually in EditorPropertyEasing

Double-clicking on the EditorPropertyEasing widget (e.g. for the Light
Attenuation parameter) shows an EditorSpinSlider to set the value manually.

Fixes #8449.

Co-authored-by: Rémi Verschelde <rverschelde@gmail.com>

ttencate committed:
Make visibility of update spinner editor-wide, hidden by default

- Make visibility of the spinner and "update always" option editor-wide,
  rather than per-project metadata.
- Add options "Show Update Spinner" and "Update Continuously" under
  Editor Settings > Interface > Editor. Both options are false by
  default.
- Rename some variables and constants to be more consistent and clearer:
  "update always" -> "update continuously", "update changes" -> "update
  when changed", "update menu" -> "update spinner", "circle" -> "update
  spinner".

Fixes #23738.

Sslaxx commented:
Do all the `state_changed` signals refer to the same kind of change? Some of them sound like they could do with a more descriptive name.

bojidar-bg committed:
Allow ColisionObject2D to get shapes from tilemaps

Fixes #4454 and likely resolves #22285

AlexHolly committed:
UndoRedo add version changed signal

added some functions to manage undo buttons

KoBeWi committed:
Add a Direction property to ParticlesMaterial

hoontee committed:
Hide command prompt launched by OS.execute

Currently, the console appears when running OS.execute in an exported project,
but not in the editor. This change prevents it from appearing in either.

Only affects console applications.

Commit:
Make easier to select keys that are at position 0 in AnimationPlayer editor

Fixes #26151.

Co-Authored-By: Michael Alexsander Silva Dias <michaelalexsander@protonmail.com>

RandomShaper committed:
Add embedded PCK option to PC platforms

The basic point is as in 2.1 (appending the PCK into the executable), but this implementation also patches a dedicated section in the ELF/PE executable so that it matches the appended data perfectly.

The usage of integer types is simplified in existing code; namely, using plain `int` for small quantities.

fire committed:
Add editor screenshot on control - f12.

goooosick committed:
fix improper uncommenting behavior in TextEditor

SonarShroom committed:
Export path may now be written as a relative path

If the target directory does not exist, it will be recursively created.

Export paths are now saved as a relative to the projects base directory

Renamed relative_to function to final_path_from_relative which takes a relative path and outputs the final path from a string that represents a directory.
Added relative_path_from_final which takes in a final path and outputs a relative path if possible. If not possible it outputs the relative path that represents the current directory.
If the target directory does not exist when exporting the project, then it is recursively created.

Removed final_path_from_relative function

Changed DirAccess into DirAccessRef for automatic object destruction

CedNaru committed:
Added a Width Curve to Line2D + UVs fix

lupoDharkael committed:
Add godot version in backtrace message

KoBeWi committed:
Smarter node placement after duplicating

vnen committed:
Keep GDScript functions in stack while yielding

This prevents GDScript functions from leaving the stack too soon when
they are resuming from yield, allowing the ones expecting it to finish
to know the caller.

Helps debugging cases when you use:
`yield(function_which_yields(), "completed")`
since now it shows the call that resumed that function.

vnen committed:
Show function name in debugger stack trace

Also show script and line when the instance is gone when resuming from
yield.

YeldhamDev committed:
Improve bookmarks

guilhermefelipecgs commented:
This pr can fix this #26408 as well.

hbina committed:
fixed an issue where there is no edited function but user double clicks on an available functions

hpvb committed:
Make sure that the resource previewer does not hang on exit

bojidar-bg committed:
Add default values to the editor help, docs, and generated RST

Also, make spacing of "=" in the editor help a bit more consistent.
Closes #16086

YeldhamDev committed:
Add 'global_rate_scale' to the AudioServer

Closes #28953.

Calinou committed:
Fix Emscripten root directory detection when building for HTML5

Recent Emscripten SDK versions seem to only include the
`BINARYEN_ROOT` variable in the Emscripten configuration file,
whereas the platform's `detect.py` only looked at `EMSCRIPTEN_ROOT`.

Calinou committed:
Make `--doctool` create directories if they don't exist

This prevents errors due to missing directories when generating
engine documentation.

This closes #17297.

nhold committed:
Add option to remove array item and button to remove typed array item

Geequlim committed:
Parse more informations for code completion

Geequlim committed:
Show icons for code completion options

m4gr3d committed:
Add XR mode selection to the Android export process.

NilsIrl committed:
Replace ` + "/" + ` with `String::file_add()`

jonri committed:
Prevent VehicleBody contact damping from exceeding the baseline value when roll influence is very small.
When contact damping exceeds 1.0, the wheels will lose grip when the vehicle is not moving.

zakstam committed:
Document MarginContainer Node

Commit:
doc: UPNP: Add example on how to open/close a port

I have added a small code sample to show how to automate the process of port forwarding (and closing after you are done).
Thanks to @mhilbrunner for explaining how to achieve t0his (see #18780)

clayjohn committed:
scale vertex lit lights by envorionment scale

zakstam committed:
Document PopupMenu node.

eligt committed:
Fix for #29810

Ensure indentation works properly in rich text
Fix formatting

akien-mga commented on scene/animation/animation_node_state_machine.cpp:
```diff
@@ -963,6 +982,10 @@ void AnimationNodeStateMachine::_bind_methods() {
 
 	ClassDB::bind_method(D_METHOD("_tree_changed"), &AnimationNodeStateMachine::_tree_changed);
+
+	ClassDB::bind_method(D_METHOD("on_playback_state_changed"), &AnimationNodeStateMachine::on_playback_state_changed);
```
This should be prefixed with `_`.

akien-mga commented:
Sorry for the late review. Commits should be squashed together and there's one change to make to the bindings. Otherwise code look fine.

zakstam committed:
Document SpinBox node

qarmin committed:
Fix TextEdit Select

clayjohn committed:
update emitting status on one-shot particles

jonri committed:
Fix updating of collision shape when the transform is set

Raphael2048 committed:
add 2d physic optimization simlilar to 3d physics

Calinou committed:
Proofread and improve the whole class reference

- Document a few more properties and methods
- Add more information to many classes
- Fix lots of typos and gramar mistakes
- Use [code] tags for parameters consistently
- Use [b] and [i] tags consistently
- Put "Warning:" and "Note:" on their own line to be more visible,
  and make them always bold
- Tweak formatting in code examples to be more readable
- Use double quotes consistently
- Add more links to third-party technologies

GlaceGwyneth committed:
Give LineEdit a custom font color when un-editable

By default the LineEdit's text when editable is unchecked had some transparency hardcoded.
This change adds a custom color to LineEdit for adjusting the font when editable is off.
Addresses issue 29814

flomar committed:
Line edit placeholders now react properly to translation changes.

Dentrax committed:
fix some crashes

NathanWarden committed:
Updated lightmap baker to dynamically calculate lightmap sizes based on surface area.

GlaceGwyneth committed:
Give TextEdit a custom color for font when read only is set

Previously there was some transparency hard coded into TextEdit when in read only mode.
This change adds a custom color for adjusting the font in read only mode.
It also applies when syntax highlighting is on.

Faless committed:
Fix PoolVector resize and subarray.

The first used to accept negative values, the second would crash if out
of bound.

Faless committed:
WebSocket module now uses wslay library.

Both client and server are supported on native builds (as usual).

SSL server is still not supported, but will soon be possible with this
new library.

The API stays the same, we just need to work out potential issues due to
this big library switch.

randall-fulton committed:
Add docstring for Object.set_deferred

KoBeWi committed:
Allow to remove 2D editor limits

akien-mga committed:
Area2D: Fix argument type of body_* signals

Those signals receive either a PhysicsBody2D or a TileMap object,
and what the emitting method checks internally is only that the
object is a Node. In theory any Node could go through these signals
if they talk directly to the PhysicsServer2D.

Also updated docs.

Fixes #27076.

Might need further (compat breaking) improvement as this API is a
bit confusing, cf. #24739.

akien-mga committed:
Fix scoped enum value reference, breaks GCC 5

Introduced in #29376 and recent compilers are fine with it, but GCC 5
complains.

Fixes #30044.

randall-fulton committed:
Call out scaling in Spatial.translate docstring

This fixes issue #26938.

njt1982 committed:
Allow '-' character in iOS reverse DNS identifiers dissallow underscores.

guilhermefelipecgs committed:
Fix Popup::popup_centered not centering at the first call

austinried committed:
Remove references to mouse stuff from OS

There is nothing in `OS` about mouse grabbing or mouse cursors, that's in `Input`, so I'm cleaning up those references.

Chaosus committed:
Extends wrapi/wrapf docs

zakstam committed:
Fix ReferenceRect node

ReferenceRect node by default was only displaying in the editor and not when the game is running.

GlaceGwyneth committed:
Fix GridMap transform problems

akien-mga committed:
doctool: Fix writing theme_item descriptions

We already had support for parsing and saving theme_item descriptions
in DocData, and displaying it in the editor, but doctool would drop the
changes as it was not writing them back to the XML.

Part of #29868.

mbrlabs committed:
Better pre-selection of search result nodes in "Create New Node" dialog.

Also search for substrings in class types, not just subsequences. If for
the current search term a substring has been found in a class type name,
prefer the substring match over the subsequence.

Fixes #26010

Zylann committed:
Document VisualServer.instances_cull_ray

YeldhamDev committed:
Fix certain items not appearing in search tree when a feature profile is set

akien-mga committed:
doc: Document EncodedObjectAsID, expose its property

qarmin committed:
Some code changed with Clang-Tidy

akien-mga committed:
Node: Fix logic of has/get_node_and_resource and document it

Also document NodePath.

akien-mga committed:
doc: Proofread and complete various nodes

All 100% completed: MainLoop, Node, Object, Path, Performance,
Reference, Resource, SceneState, SceneTree, UndoRedo.

Also fixed some en_GB occurrences as the reference spelling is en_US.

akien-mga committed:
doc: Sync classref with current source

Anutrix committed:
Fixed parser treating compound assignment the same as assignment which gave wrong argument usage count

Chaosus committed:
Shows menu when dragging connection on empty space in visual shader graph

groud committed:
Centers the 2D viewport in the editor

Commit:
Bind is_point_in_polygon in Geometry singleton

santouits committed:
Free server id pools directly.

When closing the game, we flush the command queue but after we are pushing the freeing calls of the id pool to the
command queue and they are never being run. Now we free them directly.

AgentEnder committed:
AStar get_closest_point/get_closest_position_in_segment enabled consideration

akien-mga committed:
doc: Complete and harmonize all _MAX constant descriptions

akien-mga committed:
doc: Fix scoped hyperlinks to same-class enums/methods

The class name does not need to be specified when it is the current class.

akien-mga committed:
doc: Replace all [code]CONSTANT[/code] by new [constant CONSTANT] hyperlinks

akien-mga committed:
Properly expose all Error constants to scripting languages

akien-mga committed:
doc: Don't use `GlobalScope` scope in hyperlinks, it's automatically inferred

DrMoriarty committed:
Add encrypted files support to ConfigFile

Fix #26477
Add in ConfigFile this methods:
load_encrypted(path, key)
load_encrypted_pass(path, password)
save_encrypted(path, key)
save_encrypted_pass(path, password)

anissen committed:
Setup GitHub "Sponsor" button

akien-mga committed:
DocData: Re-expose parametric setters and getters

Setters and getters have been hidden from the documentation when the matching
properties have been exposed, but some of them are parametric and require the
name or index of a given parameter to be used. So they need to be properly
documented with the type and name of the arguments they take.

For example, CPUParticles' `set_param(Parameter param, float value)`.

akien-mga committed:
Move CODEOWNERS and ISSUE_TEMPLATE.md to .github folder

Keeping CONTRIBUTING.md at the root so that contributors actually find it.

Also changed the custom FUNDING.yml link to point to godotengine.org/donate,
which also links to PayPal and other options.

Anutrix committed:
Fixed search highlight when search text not found or empty

qarmin committed:
Fix small memory leak in OS.cpp

Chaosus committed:
Auto-change file extension in EditorFileDialog when filter changes

Anutrix committed:
Fixed regression bug caused in #30095 and actually fix the issue it was supposed to fix(#26850)

Chaosus committed:
Added convertor from VisualShader to Shader

bojidar-bg committed:
Ensure --export implies --editor

Fixes #30149

qarmin committed:
Fix memory leak in Tilemaps

Chaosus committed:
Fix some bugs inside ColorPicker

Faless committed:
Properly disable SIGPIPE on OSX network sockets.

Disable SO_NOSIGPIPE socket option when avaiable.
Use MSG_NOSIGNAL send flag on systems that support it.

Faless committed:
Fix ip_unix.cpp inclusion order for OpenBSD.

Calinou committed:
Add a configuration warning when using Ignore mouse filter with tooltip

This closes #30144.

KoBeWi committed:
Clarify docs for _ready and request_ready

Commit:
Add support for creating editor icons per module

The functionality is similar to how `doc_classes` are retrieved per module.
The build system will search for custom icons path defined per module via
`get_icons_path()` method in `config.py` or default icons path.
If such paths don't exist, only the editor's own icons will be built.

Most module icons were moved from editor/icons to respective modules.

Calinou committed:
Draw the outline as part of the Label font shadow

This closes #30165.

LikeLakers2 committed:
Fix an EditorPropertyResource bug where clicking on the flat buttons would not select the property in the inspector

hbina committed:
some variables in methods.py are not used

Thus they can be safely ignored

hbina committed:
Many fallthrough switch cases now have the FALLTHROUGH macro to tell the compiler that this is intended.

hbina committed:
Pylint warning for extract.py

pylint is complaining that an "Anomalous backslash in string: '\-'"
The string constant might be missing an 'r' prefix.

hbina committed:
use vformat for duplicate strings

many strings in visual_shader_editor_plugin.cpp does not use vformat for duplicate strings making it harder to translate

Andrettin committed:
Fixed cases of unsupported pointers being used to construct variants

akien-mga committed:
doc: Sync classref with current source

akien-mga committed:
doc: Add default values to all properties

Thanks to @bojidar-bg's impressive work in #29380.

Anutrix committed:
Added descriptions to OS.get_splash_tick_msec() and OS.get_window_safe_area()

Andrettin committed:
Added support for passing functions pointers (with a class instance as the first parameter) to method bindings

bojidar-bg committed:
Remove bogus nulls from generated default values

Also, fix crash in PluginScript destructor.

akien-mga committed:
doc: Remove hardcoded default values from descriptions

They are now generated automatically by doctool.

akien-mga committed:
doc: Remove null default values that can't be determined

Applying #30187.

akien-mga committed:
CPUParticles2D: Fix gravity to be the same as in Particles2D and Area2D

This was likely a typo, 10 times the standard gravity is 98.0665
(ca. 98), not 98.8.

akien-mga committed:
Fix uninitialized variables in Line2D, JSONParseResult and TileMap

LikeLakers2 committed:
ScriptCreateDialog should emit the script_changed signal first

Commit:
Added example on how to use RandomNumberGenerator

LikeLakers2 committed:
SceneTreeDock will now only attach scripts to the selected node if the ScriptCreateDialog was opened from the SceneTreeDock

akien-mga committed:
doc: Mark PhysicsBody bounce/friction as deprecated

akien-mga committed:
doc: Enforce default values for OS properties

Otherwise we end up fetching values from the current OS instance
when running doctool, so they would change based on the system or
even simply due to changes to the system clipboard.

akien-mga committed:
doc: Complete GlobalScope documentation

Faless committed:
Update miniupnpc library to latest master

Faless committed:
Add TCP Server is_listening method

Calinou committed:
Draw the script editor's line length guideline below characters

This prevents characters from looking strange if they cross the
line length guideline.

Faless committed:
Fix HTTPRequest status checks.

HTTPRequest used to have its own `status` variable but it would never be
updated since the status of the client would be used instead.
This caused potential crashes in some edge cases.
The `status` variable is now removed, and the client status is used in
checks instead.

Calinou committed:
Tweak gamepad button names to be consistent with the documentation

Calinou committed:
Fix Nintendo buttons in the @GlobalScope documentation

These were mistakenly referred to as DualShock buttons.

qarmin committed:
Remove unnecessary code and add some error explanations

akien-mga committed:
Revert "Add godot version in backtrace message"

qarmin committed:
Remove thread to prevent memory leak

Faless committed:
Remove libwebsocket. No longer used, yay!

YeldhamDev committed:
Add "Go To" menu to the script editor and move the bookmark and breakpoint menus there

zwostein committed:
Explicitly set a channel map when initialising the Pulseaudio stream.

Ranoller committed:
Tilemap fix displaced textures and shapes and added center texture and compatibility mode

This commit fix #22989 #15249 #28206. Main problem is that tilemap displace textures in different tile origins in a strange way and doesn´t respect coincidence between texture and shapes in not uniform tiles. This issue is present in godot 3.0 and godot 3.1. To maintain compatibility are added a compatibility mode and a center texture option. Other related issues and pull request: #28896 #29487 #29519 #29961. Idications of #30204 are added

Faless committed:
CryptoCore class to access to base crypto utils.

Godot core needs MD5/SHA256/AES/Base64 which used to be provided by
separate libraries.
Since we bundle mbedtls in most cases, and we can easily only include
the needed sources if we so desire, let's use it.

To simplify library changes in the future, and better isolate header
dependencies all functions have been wrapped around inside a class in
`core/math/crypto_base.h`.

If the mbedtls module is disabled, we only bundle the needed source
files independently of the `builtin_mbedtls` option.
If the module is enabled, the `builtin_mbedtls` option works as usual.

Also remove some unused headers from StreamPeerMbedTLS which were
causing build issues.

akien-mga committed:
Update AUTHORS and DONORS list

New contributor added to AUTHORS:
@aqnuep

Thanks to all contributors and donors for making Godot possible!

akien-mga committed:
doc: Sync classref with current source

akien-mga committed:
i18n: Sync translations with Weblate

marxin committed:
Fix few GCC9 warnings:

thirdparty/assimp/include/assimp/material.inl: In member function 'aiReturn aiMaterial::Get(const char*, unsigned int, unsigned int, aiColor3D&) const':
thirdparty/assimp/include/assimp/material.inl:176:33: error: implicitly-declared 'aiColor3D& aiColor3D::operator=(const aiColor3D&)' is deprecated [-Werror=deprecated-copy]
  176 |     pOut = aiColor3D(c.r,c.g,c.b);

modules/dds/texture_loader_dds.cpp:167:50: error: comparison of integer expressions of different signedness: 'uint32_t' {aka 'unsigned int'} and 'int' [-Werror=sign-compare]
  167 |  if (format_flags & DDPF_FOURCC && format_fourcc == PF_FOURCC("DXT1")) {
      |                                                  ^
[ 28%] modules/dds/texture_loader_dds.cpp:170:57: error: comparison of integer expressions of different signedness: 'uint32_t' {aka 'unsigned int'} and 'int' [-Werror=sign-compare]
  170 |  } else if (format_flags & DDPF_FOURCC && format_fourcc == PF_FOURCC("DXT3")) {
      |                                                         ^
modules/dds/texture_loader_dds.cpp:174:57: error: comparison of integer expressions of different signedness: 'uint32_t' {aka 'unsigned int'} and 'int' [-Werror=sign-compare]
  174 |  } else if (format_flags & DDPF_FOURCC && format_fourcc == PF_FOURCC("DXT5")) {
      |                                                         ^
modules/dds/texture_loader_dds.cpp:177:57: error: comparison of integer expressions of different signedness: 'uint32_t' {aka 'unsigned int'} and 'int' [-Werror=sign-compare]
  177 |  } else if (format_flags & DDPF_FOURCC && format_fourcc == PF_FOURCC("ATI1")) {
      |                                                         ^
modules/dds/texture_loader_dds.cpp:180:57: error: comparison of integer expressions of different signedness: 'uint32_t' {aka 'unsigned int'} and 'int' [-Werror=sign-compare]
  180 |  } else if (format_flags & DDPF_FOURCC && format_fourcc == PF_FOURCC("ATI2")) {
      |                                                         ^
modules/dds/texture_loader_dds.cpp:183:57: error: comparison of integer expressions of different signedness: 'uint32_t' {aka 'unsigned int'} and 'int' [-Werror=sign-compare]
  183 |  } else if (format_flags & DDPF_FOURCC && format_fourcc == PF_FOURCC("A2XY")) {
      |                                                         ^

thirdparty/assimp/include/assimp/material.inl: In member function 'aiReturn aiMaterial::Get(const char*, unsigned int, unsigned int, aiColor3D&) const':
thirdparty/assimp/include/assimp/material.inl:176:33: error: implicitly-declared 'aiColor3D& aiColor3D::operator=(const aiColor3D&)' is deprecated [-Werror=deprecated-copy]
  176 |     pOut = aiColor3D(c.r,c.g,c.b);

marxin committed:
Use GCC 9 in travis.

akien-mga committed:
Fix TTR strings for VisualShaderNodes

It had been done wrongly in #30185, but I missed it.

akien-mga committed:
i18n: Sync translation template with current source

Faless committed:
Fix editor crash when StreamPeerSSL is unavilable.

The fix for EditorNode is a bit hacky, but the handling of the buttons
and features there is hacky too (based on enums that might not reflect
the actual state).

akien-mga committed:
ConnectionsDock: Add tooltip with signal description

Had to create a custom Tree to override `make_custom_control` for the
rich text display, similar to EditorProperty.

Fixes #30244.

Faless committed:
Add sha1 functions to string (using new CryptoCore)

bojidar-bg committed:
Fix various memory leaks and errors

Faless committed:
Add b64 to string helper in CryptoCore

m4gr3d committed:
Make the meta-data name attribute settable. This will facilitate reuse for ARCore integration.

akien-mga committed:
AppVeyor: Skip saving cache on non-master branches

Otherwise we run into situations where commits to stable branches
induce very long build times, as they have to basically build from
scratch but also invalidate the cache for future commits on the
master branch.

This commit also makes the cache folder branch-specific, but since
it's still limited to 1 GB of total cache size, we don't enable it
for non-master, as we would still run into issues with non-master
build invalidating the master cache.

Calinou committed:
Implement snapping in the CSG gizmos and 3D polygon editor

The spatial editor snap settings will now affect the CSG gizmos
and 3D polygon editor (which is used in CSGPolygon).

akien-mga committed:
SCons: Use CPPDEFINES instead of CPPFLAGS for pre-processor defines

It's the recommended way to set those, and is more portable
(automatically prepends -D for GCC/Clang and /D for MSVC).

We still use CPPFLAGS for some pre-processor flags which are not
defines.

neikeq committed:
Re-write mono module editor code in C#

Make the build system automatically build the C# Api assemblies to be shipped with the editor.
Make the editor, editor player and debug export templates use Api assemblies built with debug symbols.
Always run MSBuild to build the editor tools and Api assemblies when building Godot.
Several bugs fixed related to assembly hot reloading and restoring state.
Fix StringExtensions internal calls not being registered correctly, resulting in MissingMethodException.

neikeq committed:
Script API methods must return Ref<T> instead of Reference*

ptrcall assumes methods that return a Reference type do so with Ref<T>. Returning Reference* from a method exposed to the scripting API completely breaks ptrcalls to this method (it can be quite hard to debug!).

neikeq committed:
Fix localize_path not always working

We make sure the resource dir path ends with a trailing '/' for safety reasons, so we must make sure the path we compare it to does so as well.

Faless committed:
Godot addition to wslay, fixes MSVC build

Faless committed:
Optimize header reads (avoid extra allocs)

Calinou committed:
Implement snapping for most 3D gizmos

bojidar-bg committed:
Fix parsing of arguments in constant expressions

Fixes #8006

bojidar-bg committed:
Fix inheriting from class_name messing up constants

Fixes #29586

qarmin committed:
Fix crash with TileSet(again)

neikeq committed:
Mono: Android build and shared libraries fixes

Fix location of Mono's shared libraries.
Fix build failing if the directory 'platform/android/java/libs/{target}/{abi}' doesn't exist.

neikeq committed:
Mono: Fix Array IndexOutOfRangeException not being thrown

Calinou committed:
Remove unused icons in the default theme

DavidSichma committed:
Added force_update_cache() to RemoteTransform and RemoteTransform2D

akien-mga committed:
FileSystem dock: Improve duplicate check for directory paths

Supersedes and closes #30225.

akien-mga committed:
doc: Sync classref with current source

zakstam committed:
Adds timeout property for httprequest

codecustard committed:
Resolves Issue #30271 Reparent selection without flattening

When reparenting a multi-selection via drag and drop, the hierarchy was flattened. This change resolves that issue.

neikeq committed:
Expose ResourceImporter to the scripting API

Fixes #30127

bojidar-bg committed:
Ensure that nested YSort preserves intermediate modulate values

Fixes #28501

Calinou committed:
Use CheckBoxes in the editor instead of CheckButtons when applicable

CheckButtons should only be used if toggling them has an
immediate effect. Otherwise, CheckBoxes should be used.

LikeLakers2 committed:
SceneTreeDock will reset the ScriptCreateDialog's inheritance base type

Anutrix committed:
Got replace-all working without breaking search

Calinou committed:
Dim the editor window while it's quitting

This makes it clearer that the editor window is busy while it's quitting
(which can take a while on slower PCs).
This also makes it feel more responsive to user input.

Note that the effect is only visible if
"Interface > Editor > Quit Confirmation" is disabled in the
Editor Settings, as the quit confirmation already makes
the editor window dim while quitting.

mbrlabs committed:
Prevent GLES2 bool uniforms from having a precision type set.

When setting the default precision type for uniforms (before compiling
the shader) prevent boolean uniforms from having one set. Booleans can't
have a precision type and on some Android devices this caused a
compilation failure.

Fixes #30317

marcotc committed:
Document constants in Transform/Transform2D

Calinou committed:
Improve the CSG shape gizmo drawing

The gizmo colors now depend on the operation. Subtraction will
result in an inverted gizmo color, whereas intersection is now displayed
as white.

A solid translucent overlay is now drawn over a selected node
to make it easier to distinguish.

fire committed:
Add editor menu to open the project data "user://" location.

akien-mga committed:
doc: Sync classref with current source

Chaosus committed:
Added DEPTH_TEXTURE to visual shaders

akien-mga committed:
Revert "Add editor menu to open the project data "user://" location."

ibrahn committed:
Added release function to PoolVector::Access.

For clarity, assign-to-release idiom for PoolVector::Read/Write
replaced with a function call.
Existing uses replaced (or removed if already handled by scope)

bojidar-bg committed:
Fix parsing of enums allowing for juxtaposed identifiers

Fixes #28727

simonpuchert committed:
cpu_particles: Return uniform density spheres.

The current implementation normalizes points from a uniform distribution
on a cube. This creates a non-uniform distribution on the sphere.

LikeLakers2 committed:
Camera2D's offset now ignores the limit property

YeldhamDev committed:
Make 'size_override_stretch' a property in 'Viewport'

luiz-lima committed:
Add Weblate badge to README.md

Geequlim committed:
Fix code completion for shader editor

Geequlim committed:
Fix light area position in the selected completion

creikey committed:
Give up finding item if started at beginning

When the current item is -1, then the loop will infinitely repeat,
constantly setting i to zero and never exiting.

ohsayan committed:
Improve localization

This commit improves localization and clarifies that once removed, the file cannot be restored

BastiaanOlij committed:
Add microphone privilege settings support back into export and add privilege exist check to camera server for iOS

Kanabenki committed:
Sort selected nodes in scene tree before duplication

Xrayez commented:
Would be good if the signal provides *from* and *to* states too (either in this or future PR).
```gdscript
func _on_state_changed(from_state, to_state):
    if from_state == 'idle' and to_state == 'jump':
        leap()
```

bojidar-bg committed:
Fix editor remaining dimmed after certain actions

Fixes #30368

simonpuchert committed:
Optimize get_closest_point_to_segment*.

By combining all scalar factors we can get rid of a scalar * vector
multiplication and a square root operation, since the resulting formula
only uses the squared length.

KoBeWi committed:
Clarify better what Node.raise does

Anutrix committed:
Added filter to SceneTreeDialog

simonpuchert committed:
particles: Return uniform density spheres.

This matches the previous change for cpu_particles.

remram44 committed:
DOC: Fix spelling

See also https://github.com/godotengine/godot-docs/pull/2595

hbina committed:
Removed a redundant check in PoolVector

The set method of PoolVector<T> performs an indexing check twice.

Meriipu committed:
TreeItem: Added a setter for the disabled-property of buttons

Anutrix committed:
Removed a pair of empty braces

Calinou committed:
Add a configuration warning when using ultrawide SpotLight with shadows

This partially addresses #12028.

Calinou committed:
Hide "Max Rate Hz" if "Max Rate" is unchecked when importing WAV sample

larpon committed:
Clearify how hex_to_int expects a '0x' prefixed string. Provide a small hex_to_int example

larpon committed:
Provide documentation for String http_escape() and http_unescape() methods

qarmin committed:
Fixes minor issues found by static analyzer

Faless committed:
(Re-)Implement subprotocols in websocket server.

Faless committed:
Add TCP connect timeout.

Default timeout is 30 seconds (i.e. after 30 seconds of calling
connect_to_host if the TCP peer is not connected the connection will
error out).

This value can be configured in project settings:
`network/limits/tcp/connect_timeout_seconds`

bojidar-bg committed:
Fix some issue with TileMap's and other nodes' boundaries

Fixes #30348
Addresses a small part of #30012

neikeq committed:
Fix InvalidCastException when loading script metadata

Old value wasn't string so we have to keep that in mind

neikeq committed:
Fix C# bindings detection of abstract classes

ClassDB::can_instance was changed recently breaking our code, as the editor is not yet initialized. Check ClassInfo::creation_func directly.

neikeq committed:
Fix --generate-mono-glue bug when directory doesn't exist

DirAccess::get_full_path(path) only works if the path exists. Implement our own abspath function.

neikeq committed:
Fix Godot.Object.ToString() infinite recursion

Should not be using Variant to String conversion as that would call ToString() again

neikeq committed:
C#: Fix some crashes during assemblies reloading

neikeq committed:
Mono: Fix build errors with tools=no

Faless committed:
Add NULL-terminator the string passed to strtol.

This is actually expected by the function although it was apparently
working in GCC without the terminator, it breaks (at least some) clang
versions.

akien-mga committed:
SCons: Fix MBEDTLS_CONFIG_FILE string escape

Fixes #30431, regression from #30277.

Faless committed:
Fix WebSocketClient consuming data during hanshake

Was missing a break of the while loop on connection.
This potentially caused early data frames to be trashed.

bojidar-bg committed:
Display a count of matches when searching in the code editor

Fixes #14513

Calinou committed:
Use base `Color()` constructors instead of `Color::html()`

This results in slightly smaller binaries (-17 KB for an editor binary)
as no strings need to be allocated.

guilhermefelipecgs committed:
Fix cursor blinking in integrated GPUs

Optimization for Input::set_custom_mouse_cursor when used inside
_process function. (Avoids cursor blinking in low end devices)

Calinou committed:
Improve the node configuration warning display

- Refer to properties explicitly when possible
- When multiple warnings are returned, always separate them by one
  blank line to make them easier to distinguish
- Improve grammar and formatting

hbina committed:
Clarify what %APPDATA% expands to in Windows

Linux users should be familiar with "~" so I think it might be unnecessary.
Also changed the file separator in the description to use Windows' instead.

hbina committed:
Fix issues where translations doesn't work

The crux of the issue is a white space between % and s.
I have fixed similar problems than found in #30063

Edited by @akien-mga to fix some more occurrences on other
placeholders.

Raphael2048 committed:
Fix DynamaicFont outline does not draw using draw_char

akien-mga committed:
TreeItem: Rename new button disable method to set_button_disabled

Follow-up to #30408.

Also sync classref with current source.

akien-mga committed:
i18n: Sync translations with Weblate

akien-mga committed:
i18n: Sync translation template with current source

Chaosus committed:
Fix "not" shader function return type

Calinou committed:
Add `width` and `antialiased` parameters to CanvasItem `draw_rect()`

This also removes some duplicated editor code which is now obsoleted
by the new parameters in `draw_rect()`.

codecustard committed:
Fix issue with displaying empty tooltip hint

Closes #30448

neikeq committed:
Fix export templates ignoring assemblies in 'res://.mono/assemblies'

neikeq committed:
Make SCons check for System MSBuild before Mono's

Faless committed:
Better detect Emscripten toolchain.

Emscripten is apparently changing the variables in its config file,
causing potential breakage of our build system.

Binaries of the latest/latest-upstream releases are located in a
subfolder of BINARYEN_ROOT called emscripten.
Binaries of the other releases (e.g. sdk-1.38.31-64bit) are instead
placed under the EMSCRIPTEN_ROOT folder.

This PR checks if BINARYEN_ROOT has a subfolder called emscripten, if
that does not exists, it falls back to checking the EMSCRIPTEN_ROOT.

This way we give precedence to the new releases, given that activating
multiple releases sequentially might result in having mismatching
BINARYEN_ROOT and EMSCRIPTEN_ROOT.

bojidar-bg committed:
Add a way to copy the text of warnings shown in the editor

Closes #20662

neikeq committed:
Fix EditorNavigationMeshGenerator registered in Core API

Also added an option to output a json file with all the ClassDB registered classes and its members. This can be used to compare the API of two different builds by a simple diff.

ayuusweetfish committed:
Fix incorrect X position of line length guideline

qarmin committed:
Use reference to constant in functions

neikeq committed:
Mono: Add missing GodotTools.sln

neikeq committed:
SCons no longer passes FrameworkPathOverride to MSBuild

neikeq committed:
Fix IndexOutOfBounds in FindMsBuildToolsPathOnWindows

szamq committed:
Fix get_data layer argument when texture is TEXTURE_TYPE_2D_ARRAY

rsubtil committed:
Allow dots for class name in popup dialog

Signed-off-by: Ev1lbl0w <ricasubtil@gmail.com>

neikeq committed:
Mono: Fix null exception in GenerateGameProject

Also fix pdb files for GodotTools.*.dll assemblies not being copied to the output directory.

akien-mga committed:
doc: Sync classref with current source

neikeq committed:
Mono: Fix null dereference in EditorExportPlatformAndroid

We need to dispose the GodotSharpExport export plugin before the editor destroys EditorSettings. Otherwise, if the GC disposes it at a later time, EditorExportPlatformAndroid will be freed after EditorSettings already was, and its device polling thread will try to access the EditorSettings singleton, resulting in null dereferencing.

neikeq committed:
Mono: Fix Api HintPath and update old game projects

Fixed Api assembly references with more than one HintPath.
Also made the editor update old C# projects use the new Api assembly HintPaths.

jedStevens committed:
clang formatting and pointer fixes

exposed to and from states in signal

jedStevens committed:
squashing

jedStevens committed:
Merge branch 'state_machine_signals' of https://github.com/jedStevens/godot into state_machine_signals

jedStevens commented:
I had some serious problems trying to squash things, I'm really unsure how to do it, so I'm going to refork and apply the changes in one commit. I'm not sure what I did to screw it up so hard...
